helix-core/src/movement.rs

```diff
@@ -1,184 +1,254 @@
+    let movement = |range: Range| -> Option<Range> {
+        let after_head = (range.head + 1).skip_newlines(slice);
+        (after_head + 1).inside(slice).then(|| {
+            let new_anchor = if range.head.is_boundary(slice) {
+                after_head
+            } else {
+                range.head.skip_newlines(slice)
+            };
+            let new_head = (range.head + 1).end_of_block(slice);
+            Some(Range::new(new_anchor, new_head))
+        })?
+    };
+    (0..count).fold(range, |range, _| movement(range).unwrap_or(range))
 }
 
-pub fn move_prev_word_start(slice: RopeSlice, mut begin: usize, count: usize) -> Option<Range> {
-    let mut with_end = false;
-    let mut end = begin;
-
-    for _ in 0..count {
-        if begin == 0 {
-            return None;
-        }
-
-        let ch = slice.char(begin);
-        let prev = slice.char(begin - 1);
-
-        if categorize(ch) != categorize(prev) {
-            begin -= 1;
-        }
-
-        // return if not skip while?
-        skip_over_prev(slice, &mut begin, |ch| ch == '\n');
+pub fn move_prev_word_start(slice: RopeSlice, range: Range, count: usize) -> Range {
+    let movement = |range: Range| -> Option<Range> {
+        (range.head > 0 && range.head.inside(slice)).then(|| {
+            let new_anchor = if range.head.saturating_sub(1).is_boundary(slice) {
+                (range.head.saturating_sub(1)).backwards_skip_newlines(slice)
+            } else {
+                range.head.backwards_skip_newlines(slice)
+            };
+            let new_head = range.head.saturating_sub(1).start_of_block(slice);
+            Some(Range::new(new_anchor, new_head))
+        })?
+    };
+    (0..count).fold(range, |range, _| movement(range).unwrap_or(range))
+}
 
-        end = begin;
+pub fn move_next_word_end(slice: RopeSlice, range: Range, count: usize) -> Range {
+    let movement = |range: Range| -> Option<Range> {
+        let after_head = (range.head + 1).skip_newlines(slice);
+        (after_head + 1).inside(slice).then(|| {
+            let new_anchor = if range.head.is_boundary(slice) {
+                after_head
+            } else {
+                range.head.skip_newlines(slice)
+            };
+            let new_head = (range.head + 1).end_of_word(slice);
+            Some(Range::new(new_anchor, new_head))
+        })?
+    };
+    (0..count).fold(range, |range, _| movement(range).unwrap_or(range))
+}
 
-        with_end = skip_over_prev(slice, &mut end, char::is_whitespace);
+// ---- util ------------
 
-        // refetch
-        let ch = slice.char(end);
+/// Private helpers to help manipulate slice indices
+pub trait SliceIndexHelpers {
+    fn outside(&self, slice: RopeSlice) -> bool;
+    fn inside(&self, slice: RopeSlice) -> bool;
+    /// The next character after this belongs
+    /// to a different `Category`
+    fn is_boundary(&self, slice: RopeSlice) -> bool;
```

Comment:
```suggestion
    fn next_char_is_boundary(&self, slice: RopeSlice) -> bool;
```
I think this can make it more obvious.